Waves rollout updates (#546)
foundriesio/docs: source/reference-manual/ota/production-targets.rst

Production Targets
==================

> As a part of going to production, a user needs to define two distinct sets of devices: test versus production.

Author: Maybe, for lines 7-8, I can add a change to the next 2 lines, like below:

```
- A **production device** is defined by a special attribute in its public certificate,
+ A **production device** is defined by a presence of a special attribute in its public certificate,
  which is set at the :ref:`device registration time <ref-factory-registration-ref>`.
+ If that attribute is absent, it designates a **test device**.
```

I have no strong opinion, as the second sentence was something obvious to me.

> As a part of going to production, a user needs to define two distinct sets of devices: test versus production.

Author: https://github.com/foundriesio/docs/pull/546/commits/d772ae97260181cec96922fba6d643c474fe07c7

A **production device** is defined by a special attribute in its public certificate,
which is set at the :ref:`device registration time <ref-factory-registration-ref>`.

As noted in the :ref:`offline keys documentation<ref-offline-keys>`,
FoundriesFactory® leverages The Update Framework (TUF) to deliver Over-the-Air (OTA) Software Updates to devices.

> Production devices receive a separate copy of TUF targets metadata than test devices — **production targets**.

Author: Probably, I need to change this too:

```suggestion
Production versus test devices receive different copies of the TUF metadata:
- Test devices receive :ref:`CI targets <ref-ci-targets>`.
- Production devices receive **production targets**, managed through so-called **waves** — a FoundriesFactory way to release updates to production devices.
```

This should make it crystall clear, an eliminate some wordiness.
@angolini does it work for you?

> Production devices receive a separate copy of TUF targets metadata than test devices — **production targets**.

Author: https://github.com/foundriesio/docs/pull/546/commits/61be25cf7352bce2d1335c69ca4567bd89a25a41

> Production TUF targets must be signed by both Foundries.io™ owned and user-owned targets signing keys.

Author: I _believe_ that (combined with my previous suggestion), this could be improved too:

```suggestion
Production targets provide an increased security compared to CI targets. They must be signed by the user-owner targets signing key, in addition to the Foundries.io™ owned signing key.
```

> Production TUF targets must be signed by both Foundries.io™ owned and user-owned targets signing keys.

Author: https://github.com/foundriesio/docs/pull/546/commits/61be25cf7352bce2d1335c69ca4567bd89a25a41

Production targets are managed through so-called **waves** — a FoundriesFactory way to release updates to production devices.
Normally, when the TUF targets metadata is updated, all devices see it and can start updating themselves.

> Waves allow Factory operators to control an exact time when devices see a new version of the TUF targets metadata.

Author: These two sentences are complementary:
- the behavior described in the first sentence is always a normal one.
- the second sentence states that _waves_ allow a user to _inject_ some control into that process.

.. note::

    Production devices can only be updated to production targets.
    The opposite is also true, production targets can only be installed on production devices.
    From the TUF perspective CI versus production TUF targets comprise two isolated sets of updates.

    Within a scope of below paragraphs a term "device" always means a "production device".

.. _ref-rm-wave:

Performing a Production OTA
---------------------------

> At some point, Factory builds become stable enough to be delivered to production devices.
> Let's assume there is a CI build version 42 that is ready to be run in production.
> To start the production release process, a user would create a new wave using the below command::

Author: That statement is one of those _legacy things_ I didn't clean up.
Probably, we need to make it even less assertive e.g. I propose to remove that sentence altogether.

Alternatively, we can postulate something vague about the user release process:
```suggestion
A user should define a process to select CI builds which need to be delivered to production devices.
Let's assume a user selected a CI build version 42 as ready to be run in production.
To start the production release process, a user would create a new wave using the below command::
```

> To start the production release process, a user would create a new wave using the below command::

Author: https://github.com/foundriesio/docs/pull/546/commits/24636bd69edca7f7e465f67dcf1631902dd4d08f

  # fioctl waves init <wave-name> <target number> <tag> -k <keys.tgz>
  fioctl waves init -k /absolute/path/to/targets.only.key.tgz v2.0-update 42 production

This creates a new TUF targets role version for production devices which listen to OTA updates for the ``production`` tag.
That TUF targets role only includes a single Target from CI build (in above example, that target version is 42).

.. note::

   We recommend that a user generates :ref:`OSTree static deltas<ref-static-deltas>` before rolling out waves to devices.

Once created, a new wave can be rolled out to Factory production devices, all at once or in phases.

> There are several ways, how a wave can be rolled out:

Author: \+

> There are several ways, how a wave can be rolled out:

Author: https://github.com/foundriesio/docs/pull/546/commits/1020c463dba0603a0ac80f591152b63f6a7ceb19

- To a subset of devices in a Factory (potentially, across several device groups, and including group-less devices).
- To all devices in a Factory.

We recommend to first roll out a wave to a dedicated device group, which contains a small number of production devices.
Another good option is to roll out a wave to a small subset of devices in a bigger device group.
Let's assume a user wants to first roll out a new ``v2.0-update`` wave to a device group called ``canary``.
This can be done using the below command::

  fioctl waves rollout v2.0-update canary

A user can roll out a wave in as many phases as their workflow requires,
before making the software update generally available.
For example, a user may decide to roll out a wave to device group ``us-east-1`` in several chunks,
after enough devices in group ``canary`` were updated successfully.
To do that, a user would run the below command sequence::

  fioctl waves rollout v2.0-update us-east-1 --limit=10
  fioctl waves rollout v2.0-update us-east-1 --limit=50
  fioctl waves rollout v2.0-update us-east-1

The above command chain rolls out a wave to 10 devices in the ``us-east-1`` group,

> then to 50 more devices, and finally to all remaining devices in that group.

Author: The total will be 60; that's why I explicitly stated _50 **more** devices_.

You are a second one asking this question after @doanac. Do I need to add more clarity?

> then to 50 more devices, and finally to all remaining devices in that group.

Reviewer: I was on the fence commenting. On one hand - its clear. On the other hand, this is complicated and maybe you can't be too clear. I'd thought maybe saying "50 more devices (60 total)" might be better.

> then to 50 more devices, and finally to all remaining devices in that group.

Author: https://github.com/foundriesio/docs/pull/546/commits/f8d0034088e42707fa5fad8e4d0c376d22a79cb2

A user may also want to roll out a wave to a subset of devices in entire fleet, across several device groups.
That can be accomplished by the below command::

  fioctl waves rollout v2.0-update --limit=5

It is possible to examine a list of devices that would be updated by a rollout command, without actually performing it::

  fioctl waves rollout v2.0-update --limit=5 --dry-run --print-uuids

.. note::

    Keep in mind that the device selection is pseudo-random, and can vary from one command run to another.

> A user can then inspect and amend that list of devices, and pass it back to the rollout command.

Author: Without a comma there are too many _ands_ in this sentence to my taste.
I used a comma here intentionally, in order to disambiguate the relation between these 3 sub-sentences.

Alternatively, a user can provide their own choice of device UUIDs to update, like in the below command::

  fioctl waves rollout v2.0-update --uuids=ab8ecb00-8ed4-42ff-90b2-815b371c0f86,7a733e81-f948-43a9-a358-56f3deb5f184

Please, check the ``fioctl waves rollout --help`` command for all available options,
or look at the :ref:`Advanced Usage <ref-rm-wave-adv>` for more complex workflows.
Hopefully, they should suit your specific production release lifecycle needs.

Typically, a user would check how the wave progresses.
A user can always monitor the status of their Factory OTA updates status using the ``fioctl status`` command.

> FoundriesFactory also provides a dedicated command to monitor the wave OTA updates status — ``fioctl wave status``.

Author: https://github.com/foundriesio/docs/pull/546/commits/727b9234f7dceb2645a304b167888e5cc65aa9d4

Eventually, a user may decide that a new software release (represented by a wave) is good enough to be generally available.
In this case, wave TUF targets need to be copied into production TUF targets for a specific tag.
In our example that is accomplished by using the below command::

  fioctl waves complete v2.0-update

> Alternatively, if a wave progresses badly a user can can cancel it using below command (unless a wave is already completed)::

Author: https://github.com/foundriesio/docs/pull/546/commits/1020c463dba0603a0ac80f591152b63f6a7ceb19

  fioctl waves cancel v2.0-update

Those devices that were successfully updated to Target 42 will continue to run it.
However, other production devices will not be updated, and will continue to run the previous version.

.. note::

>    We recommend to only use production targets (of completed waves) when flashing production devices.

Author: This is one more _legacy_ which is left almost unchanged.

I thought that we have a consensus at Foundries.io that _flashing a device_ is the process when the first image is provisioned to the device at manufacturing time. Please, let me know if there are more appropriate lexics for this.

>    We recommend to only use production targets (of completed waves) when flashing production devices.

Reviewer 2: "We recommend using a production target after a completed and validated wave for flashing new production devices" 

Maybe? (if I understand it right).

Or we can add a glossary to the "flashing" word, so it's consensus through the docs also.

>    We recommend to only use production targets (of completed waves) when flashing production devices.

Author: https://github.com/foundriesio/docs/pull/546/commits/75e5bacb19109e395d6ded0604a2b77c473e7c97

.. _ref-rm-wave-adv:

> Advanced Usage

Reviewer 2: I think this section would have a better flow by using subsections for the 4(?) different examples.

Author: That was my initial thought.  Probably, I need to consider it seriously.

Author: https://github.com/foundriesio/docs/pull/546/commits/7e90a8956d38b37bd570a22d099f44b57c2e7646

--------------

The FoundriesFactory ``fioctl waves rollout`` command allows implementing various release workflows.
This section focuses on some popular scenarios, which are supported by it.

> Please, bear in mind, that this is only applicable to production devices.

Author: I think that `allows + gerund` is appropriate here.

https://www.grammaring.com/to-infinitive-or-gerund-advise-recommend-allow-permit-forbid-require

> Please, bear in mind, that this is only applicable to production devices.

Reviewer 2: It was a suggestion from the tool. I'm ok with any of your decisions.

> Please, bear in mind, that this is only applicable to production devices.

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

First, lets consider the most trivial use case — a Factory with a small device fleet.
In this case, the two most convenient ways to deliver updates to your devices in a controlled way are:

- Delivering updates via static device groups.

> - Delivering updates in dynamic chunks without device groups.

Author: In this context:
- device groups are defined statically by the user.
- when not using device groups, a list of devices to update can be defined dynamically (randomly) by the server at the rollout time.

Nevertheless, I like your suggestion, and will use it.
I also plan to improve on it. I start to dislike the _device chunk_ thing, as it looks like lacking a proper definition.

> - Delivering updates in dynamic chunks without device groups.

Reviewer 2: My main problem (and it can be only my brain fighting against the words) is that dynamically means movement; randomly is only randomly. Dynamically means (for me) that it changes, but I understand that once you execute the command, the device list (randomly generated) will not change.

> - Delivering updates in dynamic chunks without device groups.

Reviewer 2: I prefer the first one, with "generated"

> - Delivering updates in dynamic chunks without device groups.

Author: https://github.com/foundriesio/docs/pull/546/commits/f99e9d688120aef736dd66897b6e224bc7c30800

Each of the above options has its pros and cons.
Either way, the idea is to incrementally deliver the updates to your device fleet in chunks, which gradually increase in size.

> Using device groups a typical setup would look this way.

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

Assume you have a fleet of 100 devices.
We recommend to split up 2 device group out of that fleet: e.g. "canary" having 5 devices, "beta" having 20 devices.
Canary devices would usually be those that are easier to reach out to in case of any troubles during an update.
Having done that, a regular update rollout process would look like this::

    fioctl waves rollout v2.0-update --group canary
    fioctl waves rollout v2.0-update --group beta
    fioctl waves complete

.. note::

    Apparently, there must be a *wait and watch* period after each rollout command,
    before proceeding to the next one.
    Usually, a user would run the ``fioctl waves status`` command several times during that period.
    That allows a user to watch a nearly real-time picture about how the update is going on fleet-wide.

>     It is up to the user to define the acceptance criterion for going to the next phase of the rollout.

Author: Yes, I admit that your sentence order is better.

>     It is up to the user to define the acceptance criterion for going to the next phase of the rollout.

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9 and then https://github.com/foundriesio/docs/pull/546/commits/03cfc2375a05745a142bd95bc36cdb0c16b53c7e

> For the same above example, a user might opt to not use device groups, but still rollout the update in phases.

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

An equivalent way of doing this using randomized device fleet partitions may look like this::

    fioctl waves rollout v2.0-update --limit 5
    fioctl waves rollout v2.0-update --limit 20
    fioctl waves complete

Alternatively, a user may create a file containing a comma-separated list of "canary" device UUIDs.
For example, lets assume the user created a file ``canary-devices.lst`` for thse "canary" devices,
and prefers to keep the next rollout phase randomized.

> This way is even closer to the use of device groups, but does not necessitate their management::

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

    fioctl waves rollout v2.0-update --uuids @/path/to/canary-devices.lst
    fioctl waves rollout v2.0-update --limit 20
    fioctl waves complete

When using dynamic randomized device partitions for the rollout process,
Foundries.io APIs prioritize recently active devices over the offline devices.
It also filters out those devices, which already participate in a wave rollout.
That helps getting early response about how the rollout progresses,
and streamline the gradual update of your device fleet to a newer version.

Now lets look at a more complex examples of a rolling out an update to a large device fleet.
Let's assume that your Factory has 100'000 production devices to be updated within a wave.
Normally, these devices will be split into a couple of groups according to your criteria.
Let's assume your device fleet is arranged according to geographic presence,
e.g. device groups are "us-east", "us-west", "eu-emea", and "apac".
Apparently, some or all of these device groups would still contain a large amount of devices.

> So, from the safety perspective, it is risky to deliver an update to the entirety of any group like that.

Author: I've never heard of examples where `e.g.` or `i.e.` is additionally separated by a comma.

> So, from the safety perspective, it is risky to deliver an update to the entirety of any group like that.

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

> A usual practice would be to leverage the "canary" approach described above, separately to every individual device group.

Reviewer 2: ```suggestion
A usual practice would be to separately leverage the "canary" approach described above to every individual device group.
```

Author: I'd rather remove that word:
```suggestion
A usual practice would be to apply the "canary" approach (described above) to every individual device group.
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

For example, commands below would roll out a wave to the "us-east" group in 4 incremental chunks::

    fioctl waves rollout v2.0-update --group us-east --limit 5
    fioctl waves rollout v2.0-update --group us-east --limit 20
    fioctl waves rollout v2.0-update --group us-east --limit 100
    fioctl waves rollout v2.0-update --group us-east

A user can then use the same technique to roll out an update to other device groups.

An ability to specify the exact list of device UUIDs to the rollout command allows you to integrate it with your device management system.

> For example, lets assume that your "eu-emea" device group is the biggest one, containing 40'000 devices.

Reviewer 2: ```suggestion
For example, let's assume that your "eu-emea" device group is the biggest, containing 40'000 devices.
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

For that, you'd export the appropriate subsets of device UUIDs into one or more files in a Comma Separated Values (CSV) format.
We support various characters as separators: a comma, a semicolon, and all sorts of newlines and white space.
For example, lets assume a user prepared the following lists of device UUIDs::

- 4 equal partitions ``phase1.lst, phase2.lst, phase3.lst, phase4.lst``, containing 10'000 devices each.
- a partition ``canary.lst``, containing 20 carefully pre-selected "canary" devices, that may intersect with the above partitions.

That would allow you to roll out an update to the device group "eu-emea" in an even more controlled way::

    fioctl waves rollout v2.0-update --group eu-emea --uuids @/path/to/canary.lst
    fioctl waves rollout v2.0-update --group eu-emea --limit 100 --uuids @/path/to/phase1.lst
    fioctl waves rollout v2.0-update --group eu-emea --limit 100 --uuids @/path/to/phase2.lst
    fioctl waves rollout v2.0-update --group eu-emea --limit 100 --uuids @/path/to/phase3.lst
    fioctl waves rollout v2.0-update --group eu-emea --limit 100 --uuids @/path/to/phase4.lst
    fioctl waves rollout v2.0-update --group eu-emea

> Above commands roll out to all "canary" devices, then to 100 random devices in each "phase",
> and finally to the remainder of the device group.

Reviewer 2: ```suggestion
The above commands roll out to all "canary" devices, then to 100 random devices in each "phase,"
and finally, to the remainder of the device group.
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

.. note::

>     At Foundries.io we care a lot about the speed of our APIs and scaling to large device fleets.
>     That strategy binds us to define certain limits for specific device management operations.
>     One such limit is that a user cannot pass more than 10'000 device UUIDs to a single rollout command.
>     That constrain also implies that the ``--limit`` argument does not accept a value bigger than 10'000.
>     It is still possible to pass more than 10'000 device UUIDs using several rollout commands.
>     Also, a user can roll out to the entire device group.

Reviewer 2: ```suggestion
    At Foundries.io, we care a lot about the speed of our APIs and scaling to large device fleets.
    That strategy binds us to define certain limits for specific device management operations.
    One such limit is that users cannot pass more than 10'000 device UUIDs to a single rollout command.
    That constraint also implies that the ``--limit`` argument does not accept a value bigger than 10'000.
    It is still possible to pass more than 10'000 device UUIDs using several rollout commands.
    Also, a user can roll out to the entire device group.
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

>     When rolling out to a subset of devices using ``--limit`` argument,
>     the "randomized" sample will exclude those devices that were already updated to a wave version.
>     It also tries to exclude those devices that were staged for update
>     (included into the device UUID list) in previous rollout command, but not yet updated to a wave version.

Reviewer 2: ```suggestion
    When rolling out to a subset of devices using ``--limit`` argument,
    the "randomized" sample will exclude devices already updated to a wave version.
    It also tries to exclude those devices that were staged for update
    (included in the device UUID list) in the previous rollout command, but not yet updated to a wave version.
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

>     A precision of the latter criteria drops if previous rollout commands to the same group provided more than 10'000 device UUIDs in total.

Reviewer 2: ```suggestion
    Precision of the latter criteria drops if previous rollout commands to the same group provided more than 10'000 device UUIDs in total.
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

    In particular, the same (not yet updated) device can be selected for the rollout several times.
    That precision loss allows us to keep the decision making speed reasonable,
    regardless of the number of devices in your Factory, theoretically scaling to infinity.
    A user can restore a lossless precision by specifying both ``--uuids`` and ``--limit`` arguments, as described in an example above.

> The same techniques as described above can be applied without using the ``--group`` argument.
> In this case, the rollout command will be applied to a subset of the entire device fleet.
> For example, below command rolls out a wave to 5'000 devices in a ``pre-selected.lst`` file,
> across the entire fleet in 4 incremental chunks::

Reviewer 2: ```suggestion
The techniques described above can be applied without using the ``-group`` argument.
In this case, the rollout command will be applied to a subset of the entire device fleet.
For example, the below command rolls out a wave to 5'000 devices in a ``pre-selected.lst`` file
across the entire fleet in 4 incremental chunks::
```

Author: https://github.com/foundriesio/docs/pull/546/commits/fa24db2fb700d81f41fcea673e1e0dd05058dde9

==================

As noted in the :ref:`offline keys documentation<ref-offline-keys>`,
FoundriesFactory® leverages The Update Framework (TUF) to deliver Over-the-Air (OTA) Software Updates to devices.

> By design, production devices receive a slightly different copy of TUF targets metadata than test devices.

Reviewer 3: `a slightly different` - a bit vague statement.

Author: Yes, indeed.

I paid some attention to the old text; but should pay more.
Simply removing a wordy `slightly` adjective seems sane.

Production TUF targets must be signed by both Foundries.io™ owned and user-owned targets signing keys.

Production targets are managed through so-called **waves** — a FoundriesFactory way to release production updates.
Normally, when the TUF targets metadata is updated, all devices see it and can start updating themselves.
Waves allow Factory operators to control an exact time when devices see a new version of the TUF targets metadata.

.. note::

>     Production devices can only be updated to production targets.

Reviewer 3: I would make explicit definition of "production targets" term  and "production devices" or provides a reference to such the definition if it's already exists in the parts of the overall user doc.

IMO, I should be something like "production targets are targets that...".

Reviewer 3: I mean that the term "production targets" is  implicitly defined above by "Production TUF targets must be signed..." and "Production targets are managed through...", but I understand it because I know what it is, not sure a new factory user will fully get it.

Maybe, "production targets are targets that are signed by ... and managed through ...".
Or, maybe avoiding a passive voice at all. E.g. "By default, TUF targets are not available for production devices. To turn a TUF target to a production, a user must signed it with two keys, ..... A user can accomplish it through mechanism of waves..

Author: What you say makes sense.

There is a sort of explanation how production devices & targets work in the [Offline Factory TUF Keys](https://docs.foundries.io/latest/reference-manual/security/offline-keys.html#how-to-rotate-offline-tuf-targets-key) manual. But it is far from qualifying as a definition. 

This place seems like a logical place to define such things. So, I'll try to make something edible, leveraging the examples you gave.

Author: I added a definition of a production device.
A definition of production targets is already there, I just clarified and highlighted it.

https://github.com/foundriesio/docs/pull/546/commits/3aa63cd5632fe61837a54d6ce4281ffd1b7cb694

    The opposite is also true, production targets can only be installed on production devices.
    From the TUF perspective these are two isolated sets of updates.

    Within a scope of below paragraphs a term "device" always means a "production device".

.. _ref-rm-wave:

Performing a Production OTA
---------------------------

At some point, Factory builds become stable enough to be delivered to production devices.
Let's assume there is a CI build version 42 that is ready to be run in production.
To start the production release process, a user would create a new wave using the below command::

  # fioctl waves init <wave-name> <target number> <tag> -k <keys.tgz>
  fioctl waves init -k /absolute/path/to/targets.only.key.tgz v2.0-update 42 production

> This creates a new TUF targets role for production devices which listen to OTA updates for the ``production`` tag.

Reviewer 3: I sounds like "this creates the role". The targets role is not created, metadata of a TUF targets role are created?

Author: You're right. It should read as `This creates a new TUF targets role version`.

Author: https://github.com/foundriesio/docs/pull/546/commits/535b7f7544b92c2291a761024bf29471ff99d2ac

That TUF targets role only includes a single Target from CI build (in above example, that target version is 42).

.. note::

>    A user may want to generate :ref:`OSTree static deltas<ref-static-deltas>` before rolling out waves to devices.

Reviewer 3: Maybe "we strongly advise to generate..."?

Author: I may paraphrase this to `We recommend`.

I prefer to stick to the [IETF definition of key words for RFCs](https://datatracker.ietf.org/doc/html/rfc2119) when expressing imperatives. For example, if something cannot be done a different way, use words _YOU MUST or WE REQUIRE_; and when we express our preference to do something in a specific way (even if we insist on that), use words _YOU SHOULD or WE RECOMMEND_".

Author: @mike-sul you choose, please.

Reviewer 3: I opt for "recommend".

Author: https://github.com/foundriesio/docs/pull/546/commits/607574f87d4b3e4ffdaf5cd25e8c22c210c68c7d

An equivalent way of doing this using randomized device fleet partitions may look like this::

    fioctl waves rollout v2.0-update --limit 5
    fioctl waves rollout v2.0-update --limit 20
    fioctl waves complete

Alternatively, a user may create a file containing a comma-separated list of "canary" device UUIDs.
For example, lets assume the user created a file ``canary-devices.lst`` for thse "canary" devices,
and prefers to keep the next rollout phase randomized.
This way is even closer to the use of device groups, but does not necessitate their management::

    fioctl waves rollout v2.0-update --uuids @/path/to/canary-devices.lst
    fioctl waves rollout v2.0-update --limit 20
    fioctl waves complete

When using dynamic randomized device partitions for the rollout process,
Foundries.io APIs prioritize recently active devices over the offline devices.
It also filters out those devices, which already participate in a wave rollout.
That helps getting early response about how the rollout progresses,
and streamline the gradual update of your device fleet to a newer version.

Now lets look at a more complex examples of a rolling out an update to a large device fleet.

> Let's assume that your Factory has 100'000 production devices to be updated within a wave.

Reviewer: i'm not sure we have guidance on the thousands separator. The apostrophe catches my eye. 

@kprosise will know how we do this

Author: IIRC the classical English notation for large numbers is to use a comma to separate orders (e.g. `1,000,000`). However, I usually prefer a guidance for globalization, adviced by many sources including Wikipedia:
https://en.wikipedia.org/wiki/English_numerals#Very_large_numbers

> Often, large numbers are written with (preferably non-breaking half-spaces or thin spaces separating the thousands (and, sometimes, with normal spaces or apostrophes) instead of commas — to ensure that confusion is not caused in countries where a decimal comma is used.

As regular spaces render badly when wrapping, and half-spaces are hard to input in Vim, I prefer to use the apostroph out of widespread choices.

==================

As noted in the :ref:`offline keys documentation<ref-offline-keys>`,
FoundriesFactory® leverages The Update Framework (TUF) to deliver Over-the-Air (OTA) Software Updates to devices.
By design, production devices receive a slightly different copy of TUF targets metadata than test devices.
Production TUF targets must be signed by both Foundries.io™ owned and user-owned targets signing keys.

Production targets are managed through so-called **waves** — a FoundriesFactory way to release production updates.
Normally, when the TUF targets metadata is updated, all devices see it and can start updating themselves.
Waves allow Factory operators to control an exact time when devices see a new version of the TUF targets metadata.

.. _ref-rm-wave:

Performing a Production OTA
---------------------------

At some point, Factory builds become stable enough to be delivered to production devices.

> Lets assume there is a CI build version 42 that is ready to be run in production.

Author: +

> Lets assume there is a CI build version 42 that is ready to be run in production.

Author: Fixed in force-push

To start the production release process, a user would create a new wave using the below command::

  # fioctl waves init <wave-name> <target number> <tag> -k <keys.tgz>
  fioctl waves init -k /absolute/path/to/targets.only.key.tgz v2.0-update 42 production

This creates a new TUF targets role for production devices which listen to OTA updates for the ``production`` tag.
That TUF targets role only includes a single Target from CI build (in above example, that target version is 42).

.. note::

   A user may want to generate :ref:`OSTree static deltas<ref-static-deltas>` before rolling out waves to devices.

Once created, a new wave can be rolled out to Factory production devices, all at once or in phases.
There are several ways, how a wave can be rolled out:

> - To a subset of devices in a specific device group.
> - To all devices in a specific device group.
> - To a subset of devices in a Factory (potentially, across several device groups, and including group-less devices).
> - To all devices in a Factory.

Reviewer 4: Should we mention again that these are production devices only or is this already implied enough?
```suggestion
- To a subset of production devices in a specific device group.
- To all production devices in a specific device group.
- To a subset of production devices in a Factory (potentially, across several device groups, and including group-less devices).
- To all production devices in a Factory.
```

Author: I think I need to stress in a note somewhere at the page top that this entire document only applies to production devices.
Some of the sentences here are already rather wordy, including the "production" everywhere would make it even worse.

@vanmaegima what do you think?

Reviewer 4: Agreed, a note seems like a good idea.

Author: Added a note in force-push.

We recommend to first roll out a wave to a dedicated device group, which contains a small number of production devices.
Another good option is to roll out a wave to a small subset of devices in a bigger device group.

> Lets assume a user wants to first roll out a new ``v2.0-update`` wave to a device group called ``canary``.

Reviewer 4: ```suggestion
Let's assume a user wants to first roll out a new ``v2.0-update`` wave to a device group called ``canary``.
```

Author: Fixed in force-push

This can be done using the below command::

  fioctl waves rollout v2.0-update canary

A user can roll out a wave in as many phases as their workflow requires,

> before making the software update generally availabile.

Reviewer 4: ```suggestion
before making the software update generally available.
```

Author: Fixed in force-push

For example, a user may decide to roll out a wave to device group ``us-east-1`` in several chunks,
after enough devices in group ``canary`` were updated successfully.
To do that, a user would run the below command sequence::

  fioctl waves rollout v2.0-update us-east-1 --limit=10
  fioctl waves rollout v2.0-update us-east-1 --limit=50
  fioctl waves rollout v2.0-update us-east-1

The above command chain rolls out a wave to 10 devices in the ``us-east-1`` group,

> then to 50 devices, and finally to all remaining devices in that group.

Reviewer: your language is accurate "50" not "50 more". I still wonder if we should try and find a more explicit way of saying the limit is the *total*

Author: fixed in force-push

    Keep in mind that the device selection is pseudo-random, and can vary from one command run to another.

A user can then inspect and amend that list of devices, and pass it back to the rollout command.
Alternatively, a user can provide their own choice of device UUIDs to update, like in the below command::

  fioctl waves rollout v2.0-update --uuids=ab8ecb00-8ed4-42ff-90b2-815b371c0f86,7a733e81-f948-43a9-a358-56f3deb5f184

Please, check the ``fioctl waves rollout --help`` command for all available options.
Hopefully, they should suit your specific production release lifecycle needs.

Typically, a user would check how the wave progresses.
A user can always monitor the status of their Factory OTA updates status using the ``fioctl status`` command.
FoundriesFactory also provides a dedicated command to monitor the wave OTA updates status — ``fioctl wave status``.

Eventually, a user may decide that a new software release (represented by a wave) is good enough to be generally available.
In this case, wave TUF targets need to be copied into production TUF targets for a specific tag.
In our example that is accomplished by using the below command::

  fioctl waves complete v2.0-update

> Alternatively, if a wave progresses badly a user can can cancel it using below command::

Reviewer 4: I believe there's no way to cancel a wave after it's been completed, correct? Just to make this clearer:
```suggestion
Alternatively, if a wave progresses badly during the initial tests a user can can cancel it using below command::
```

Author: Maybe this?

```suggestion
Alternatively, if a wave progresses badly a user can can cancel it using below command (unless a wave is already completed)::
```

Author: Fixed using my suggestion in force-push.

Lets assume your device fleet is split up according to geographic presence,
e.g. device groups are "us-east", "us-west", "eu-emea", and "apac".
Apparently, some or all of these device groups would still contain a large amount of devices.
So, from the safety perspective, it is risky to deliver an update to the entirety of any group like that.

A usual practice would be to leverage the "canary" approach described above, separately to every individual device group.
For example, commands below would roll out a wave to the "us-east" group in 4 incremental chunks::

    fioctl waves rollout v2.0-update --group us-east --limit 5
    fioctl waves rollout v2.0-update --group us-east --limit 20
    fioctl waves rollout v2.0-update --group us-east --limit 100
    fioctl waves rollout v2.0-update --group us-east

A user can then use the same technique to roll out an update to other device groups.

An ability to specify the exact list of device UUIDs to the rollout command allows you to integrate it with your device management system.
For example, lets assume that your "eu-emea" device group is the biggest one, containing 40'000 devices.

> You might use your device management system to split that fleet into several partitions:

Author: Fixed in force-push

    fioctl waves rollout v2.0-update --group eu-emea

Above commands roll out to all "canary" devices, then to 100 random devices in each "phase",
and finally to the remainder of the device group.

.. note::

    At Foundries.io we care a lot about the speed of our APIs and scaling to large device fleets.
    That strategy binds us to define certain limits for specific device management operations.
    One of such limits is that a user cannot pass more than 10'000 device UUIDs to a single rollout command.
    That constrain also implies that the ``--limit`` argument does not accept a value bigger than 10'000.
    Furthermore, if more than 10'000 devices were selected for the rollout in the device group,
    there is a loss in precision of selecting "randomized" device subgroups.
    In particular, the same (not yet updated) device can be selected for the rollout several times.
    That loss in precision allows us to keep the decision making speed reasonable,
    regardless of the total number of devices in your Factory, theoretically scaling to infinity.
    Specifying both ``--uuids`` and ``--limit`` arguments as described above,

>     you can still achieve the exact device selection process.

Author: @kprosise @vanmaegima I'm not quite fond of this paragraph. I'd be grateful if you could help me to make it neater.

@doanac I guess you need to pay attention to this too. Do we would to expose this limitation at all?

Reviewer 4: I guess we can add a note saying: limit is 10,000 devices per rollout, and have this whole text in our Confluence page about Known Issues/Limitations.
If we start to get many customer questions about this, then we can revert and put something in the docs.
What do you think?

Author: I'll try to come up with a better version tomorrow, using a fresh mind.

Reviewer: we shouldn't hide important details - having it as a note for support just means time support will have to spend on a ticket
